samples/nrf9160/serial_lte_modem/src/tcpip_proxy/slm_at_tcp_proxy.c

```diff
@@ -0,0 +1,679 @@
+K_TIMER_DEFINE(conn_timer, NULL, NULL);
+
+static struct sockaddr_in remote;
+static struct tcp_proxy_t {
+	int sock; /* Socket descriptor. */
+	int sock_peer; /* Socket descriptor for peer. */
+	int role; /* Client or Server proxy */
+} proxy;
+
+/* global functions defined in different files */
+void rsp_send(const u8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct at_param_list at_param_list;
+extern struct modem_param_info modem_param;
+extern char rsp_buf[CONFIG_AT_CMD_RESPONSE_MAX_LEN];
+
+/** forward declaration of cmd handlers **/
+static void tcp_thread_func(void *p1, void *p2, void *p3);
+
+static int do_tcp_server_start(u16_t port, int sec_tag)
+{
+	int ret = 0;
+	struct sockaddr_in local;
+	int addr_len;
+
+	/* Open socket */
+	if (sec_tag == INVALID_SEC_TAG) {
+		proxy.sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	} else {
+		sprintf(rsp_buf,
+			"#XTCPSVR: TLS Server not supported\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		return -ENOTSUP;
+	}
+	if (proxy.sock < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		sprintf(rsp_buf, "#XTCPSVR: %d\r\n", -errno);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		return -errno;
+	}
+
+	/* Bind to local port */
+	local.sin_family = AF_INET;
+	local.sin_port = htons(port);
+	ret = modem_info_params_get(&modem_param);
+	if (ret) {
+		LOG_ERR("Unable to obtain modem parameters (%d)", ret);
+		close(proxy.sock);
+		return -1;
+	}
+	addr_len = strlen(modem_param.network.ip_address.value_string);
+	if (addr_len == 0) {
+		LOG_ERR("LTE not connected yet");
+		close(proxy.sock);
+		return -1;
+	}
+	if (!check_for_ipv4(modem_param.network.ip_address.value_string,
+			addr_len)) {
+		LOG_ERR("Invalid local address");
+		close(proxy.sock);
+		return -1;
+	}
+	if (inet_pton(AF_INET, modem_param.network.ip_address.value_string,
+		&local.sin_addr) != 1) {
+		LOG_ERR("Parse local IP address failed: %d", -errno);
+		close(proxy.sock);
+		return -EINVAL;
+	}
+
+	ret = bind(proxy.sock, (struct sockaddr *)&local,
+		 sizeof(struct sockaddr_in));
+	if (ret) {
+		LOG_ERR("bind() failed: %d", -errno);
+		sprintf(rsp_buf, "#XTCPSVR: %d\r\n", -errno);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		close(proxy.sock);
+		return -errno;
+	}
+
+	/* Enable listen */
+	ret = listen(proxy.sock, 1);
+	if (ret < 0) {
+		LOG_ERR("listen() failed: %d", -errno);
+		sprintf(rsp_buf, "#XTCPSVR: %d\r\n", -errno);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		close(proxy.sock);
+		return -errno;
+	}
+
+	tcp_thread_id = k_thread_create(&tcp_thread, tcp_thread_stack,
+			K_THREAD_STACK_SIZEOF(tcp_thread_stack),
+			tcp_thread_func, NULL, NULL, NULL,
+			THREAD_PRIORITY, K_USER, K_NO_WAIT);
+
+	proxy.role = AT_TCP_ROLE_SERVER;
+	sprintf(rsp_buf, "#XTCPSVR: %d started\r\n", proxy.sock);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	return ret;
+}
+
+static int do_tcp_server_stop(int error)
+{
+	int ret = 0;
+
+	if (proxy.sock > 0) {
+		k_timer_stop(&conn_timer);
+		k_thread_abort(tcp_thread_id);
```

Comment:
Both accept() and recv() are blocking and difficult to break. Once accept() is called, I don't have a good way to cancel it. I had thought about a grace way of terminating the thread by semaphore.  Due to the blocking feature I gave up and used the brutal abort instead.

"If no pending connections are present on the queue, and the socket is not marked as nonblocking, accept() blocks the caller until a connection is present.  If the socket is marked nonblocking and no pending connections are present on the queue, accept() fails with the error EAGAIN or EWOULDBLOCK" maybe nonblocking Socket in the future?

Reply:
Well, that's a tricky topic I agree. Ideally we should use non-blocking calls, and we should monitor sockets with `poll`. Even more ideally, we would need some event notification mechanism so we could "unblock" the poll when we need to exit the thread (something like `eventfd`) but I have no hopes for that, so we'd need to rely on timeouts (possibly bsdlib could allow to close a socket from a different thread to generate an event on a socket, but this would need to be verified).

But as I said, these are complex changes which likely need some investigation first. So if you confirm that the current solution works, I guess we can leave it as it is now.